test/integration/daemonset/daemonset_test.go

```diff
@@ -393,3 +404,92 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
+	podSelector := labels.Set(labelMap).AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
+	if err != nil {
+		t.Fatalf("failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	return pods
+}
+
+func updatePodStatus(t *testing.T, podClient typedv1.PodInterface, podName string, updateStatusFunc func(*v1.Pod)) *v1.Pod {
+	var pod *v1.Pod
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(podName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		updateStatusFunc(newPod)
+		pod, err = podClient.UpdateStatus(newPod)
+		return err
+	}); err != nil {
+		t.Fatalf("failed to update status of pod %q: %v", pod.Name, err)
+	}
+	return pod
+}
+
+// waitForAllPodsPending waits for all pods to be in pending phase
+func waitForAllPodsPending(podClient corev1typed.PodInterface, t *testing.T) {
+	if err := wait.Poll(pollInterval, pollTimeout, func() (bool, error) {
+		pods := getPods(t, podClient, testLabels())
+		for _, pod := range pods.Items {
+			if pod.Status.Phase != v1.PodPending {
+				return false, nil
+			}
+		}
+		return true, nil
+	}); err != nil {
+		t.Fatalf("failed waiting for all pods to be in pending phase: %v", err)
+	}
+}
+
+// A DaemonSet should retry creating failed daemon pods
+func TestRetryCreatingFailedPod(t *testing.T) {
+	for _, strategy := range updateStrategies() {
+		server, closeFn, dc, informers, clientset := setup(t)
+		defer closeFn()
+		ns := framework.CreateTestingNamespace("retry-creating-failed-pod-test", server, t)
+		defer framework.DeleteTestingNamespace(ns, server, t)
+
+		dsClient := clientset.ExtensionsV1beta1().DaemonSets(ns.Name)
+		podClient := clientset.CoreV1().Pods(ns.Name)
+		nodeClient := clientset.CoreV1().Nodes()
+		podInformer := informers.Core().V1().Pods().Informer()
+		stopCh := make(chan struct{})
+		defer close(stopCh)
+		informers.Start(stopCh)
+		go dc.Run(5, stopCh)
+
+		dsName := "daemonset"
+		ds := newDaemonSet(dsName, ns.Name)
+		ds.Spec.UpdateStrategy = *strategy
+		ds, err := dsClient.Create(ds)
+		if err != nil {
+			t.Fatalf("failed to create daemonset %q: %v", dsName, err)
+		}
+
+		nodeName := "single-node"
+		node := newNode(nodeName, nil)
+		_, err = nodeClient.Create(node)
+		if err != nil {
+			t.Fatalf("failed to create node %q: %v", nodeName, err)
+		}
+
+		// Validate everything works correctly, include marking pods
+		// as ready and ensure they launch on every node
+		validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
+		validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 1, t)
+
+		// Set the pod's phase to "Failed"
+		pods := getPods(t, podClient, testLabels())
+		updatePodStatus(t, podClient, pods.Items[0].Name, func(pod *v1.Pod) {
+			pod.Status.Phase = v1.PodFailed
+		})
+
+		// Wait for the failed pod to be created again and in pending phase
+		waitForAllPodsPending(podClient, t)
```

Comment:
Why stop when it's pending? Doesn't it go to running?

Reply:
The pods will not get ready if we do not explicitly mark it, so they will stay at pending phase. Am I right?